subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -58,39 +60,68 @@
-        Path watchableHierarchyPath = watchableHierarchy.toPath().toAbsolutePath();
-        String watchableHierarchyPathString = watchableHierarchyPath.toString();
-        if (!watchFilter.test(watchableHierarchyPathString)) {
+    public void registerWatchableHierarchy(File watchableHierarchy, File probeFile, SnapshotHierarchy root) {
+        String watchableHierarchyPath = watchableHierarchy.getAbsolutePath();
+        if (!watchFilter.test(watchableHierarchyPath)) {
             throw new IllegalStateException(String.format(
                 "Unable to watch directory '%s' since it is within Gradle's caches",
-                watchableHierarchyPathString
+                watchableHierarchyPath
             ));
         }
-        if (unsupportedHierarchies.contains(watchableHierarchyPathString)) {
+        if (unsupportedHierarchies.contains(watchableHierarchyPath)) {
             LOGGER.info("Not watching {} since the file system is not supported", watchableHierarchy);
             return;
         }
-        if (!watchableHierarchies.contains(watchableHierarchyPathString)) {
-            checkThatNothingExistsInNewWatchableHierarchy(watchableHierarchyPathString, root);
-            recentlyUsedHierarchies.addFirst(watchableHierarchyPath);
+        if (!watchableHierarchies.contains(watchableHierarchyPath)) {
+            checkThatNothingExistsInNewWatchableHierarchy(watchableHierarchyPath, root);
+            recentlyUsedHierarchies.addFirst(watchableHierarchy);
             watchableHierarchies = watchableHierarchies.plus(watchableHierarchy);
         } else {
-            recentlyUsedHierarchies.remove(watchableHierarchyPath);
-            recentlyUsedHierarchies.addFirst(watchableHierarchyPath);
+            recentlyUsedHierarchies.remove(watchableHierarchy);
+            recentlyUsedHierarchies.addFirst(watchableHierarchy);
         }
+        WatchProbe watchProbe = new WatchProbe(watchableHierarchy, probeFile);
+        watchProbesByHierarchy.put(watchableHierarchy, watchProbe);
+        watchProbesByPath.put(probeFile.getAbsolutePath(), watchProbe);
         LOGGER.info("Now considering {} as hierarchies to watch", recentlyUsedHierarchies);
     }
 
+    public void armWatchProbe(File watchableHierarchy) {
+        WatchProbe probe = watchProbesByHierarchy.get(watchableHierarchy);
+        if (probe != null) {
+            try {
+                probe.arm();
+            } catch (IOException e) {
+                // TODO Should this be debug instead?
+                LOGGER.warn("Could not arm watch probe for hierarchy {}", watchableHierarchy, e);
+            }
+        } else {
+            // TODO Should this be debug instead?
+            LOGGER.warn("Did not find watchable hierarchy to probe: {}", watchableHierarchy);
+        }
+    }
+
+    /**
+     * Triggers a watch probe at the given location if one exists.
+     */
+    public void triggerWatchProbes(String path) {
```

Comment:
```suggestion
    public void triggerWatchProbe(String path) {
```

I think this should be singular.

Reply:
Fixed.